public/dorequest.php

```diff
@@ -114,6 +155,25 @@ function DoRequestError(string $error, ?int $status = 200, ?string $code = null)
     return DoRequestError("You do not have permission to do that.", 403, 'access_denied');
 }
 
+/*
+ * It is possible for some calls to be made on behalf of another user.
+ * This generally happens with a "Standalone" integration.
+ * NOTE: "syncachievements" is not included here because it accepts an array of
+ * achievement ID values. It doesn't use the generic `delegateUserAction()` function.
+ */
+$allowsGenericDelegation = [
+    "awardachievement",
+    "ping",
+    "startsession",
+];
+if (in_array($requestType, $allowsGenericDelegation)) {
+    try {
+        [$user, $username] = delegateUserAction($user, $delegateTo, $gameID, $achievementID);
```

Comment:
Since `delegateUserAction` is only called here, why not just embed it?
```
if (in_array($requestType, $allowsGenericDelegation)) {
    $foundTargetUser = User::firstWhere('User', $targetUsername);
    if (!$foundTargetUser) {
        return DoRequestError("The target user couldn't be found.", 403, 'access_denied');
    }
    ...
```

That would also eliminate the need to raise and catch exceptions.

Reply:
Good idea -- originally I duplicated the logic in all the cases, then I lifted it up into the function to be a bit more elegant. Direct embedding is another good step beyond that. Done in latest.